tests/go/go-kafka-serde/client.go

```diff
@@ -267,13 +253,114 @@ func (tc *TestClient) RunTest() error {
 			}
 
 			numConsumed++
-		case kafka.Error:
-			util.Die("Failure during consumption: %v", e)
+		case *kafka.Error:
+			return e
 		default:
 			log.Debugf("Ignoring unknown message: %v", e)
 		}
 	}
 
+	log.Debugf("Consumed %v messages", numConsumed)
+
+	if numConsumed != tc.count {
+		util.Die("Did not consume expected number of messages: %v != %v", numConsumed, tc.count)
+	}
+
+	return nil
+}
+
+func (tc *TestClient) produceMessages() error {
+	numDelivered := 0
+
+	// Create a context with a cancel function
+	produceCheckContext, cancelProduceCheck := context.WithCancel(context.Background())
+	// Create an error group that will be used to monitor for cancellation and
+	// message production events from franz-go producer
+	produceCheckGroup, produceCheckContext := errgroup.WithContext(produceCheckContext)
+	// Create a timeout context
+	timeoutContext, timeoutCancel := context.WithCancel(produceCheckContext)
+
+	monitorProduceFn := func(ctx context.Context) func() error {
+		return func() error {
+			for {
+				select {
+				case <-ctx.Done():
+					log.Debugf("Monitor Produce received done signal: %v", ctx.Err())
+					return nil
+				case e := <-tc.producer.Events():
+					switch ev := e.(type) {
+					case *kafka.Message:
+						m := ev
+
+						if m.TopicPartition.Error != nil {
+							log.Errorf("Failed to produce to topic %v: %v", tc.topic, m.TopicPartition.Error)
+						} else {
+							numDelivered++
+
+							if numDelivered == tc.count {
+								log.Infof("Produced expected number of messages")
+								timeoutCancel()
```

Comment:
Let's defer this cancel at the function start so we don't forget to do it in error cases.